src/dagmc/thread_manager.cpp

```diff
@@ -0,0 +1,78 @@
+#include <iostream>
+#include "thread_manager.hpp"
+
+// constructor
+DagThreadManager::DagThreadManager(int thread_count, moab::Interface* MBI) {
+  // if the moab pointer is not null point to it
+  if (MBI != NULL) {
+    MOAB = MBI;
+  } else {
+    // make new moab instance to share all DAGMC data
+    MOAB = new moab::Core();
+  }
+
+  // make a new GTT for Thread manager
+  GTT = new moab::GeomTopoTool(MOAB,false);
+  
+  // number of threads
+  set_num_threads(thread_count);
+
+  // set storage for master
+  dagmc_instances.push_back(new moab::DagMC(GTT));
+  dagmc_rayhistories.push_back(new DagMCRayState());
+
+  // setup threads
+  setup_child_threads();
+}
+
+
+DagThreadManager::DagThreadManager(moab::Interface* MBI) {
+  // if the moab pointer is not null point to it
+  if (MBI != NULL) {
+    MOAB = MBI;
+  } else {
+    // make new moab instance to share all DAGMC data
+    MOAB = new moab::Core();
+  }
+
+  // make a new GTT
+  GTT = new moab::GeomTopoTool(MOAB,false);
+  
+  // set storage for master
+  dagmc_instances.push_back(new moab::DagMC(GTT));
+  dagmc_rayhistories.push_back(new DagMCRayState());
+}
+
+// destructor
+DagThreadManager::~DagThreadManager() {
+  for (int i = 0 ; i < num_threads ; i++) {
+    // delete each dagmc instance
+    delete dagmc_instances[i];
+    delete dagmc_rayhistories[i];
+  }
+}
+
+// set the number of threads
+void DagThreadManager::set_num_threads(int thread_count) {
```

Comment:
I was a little confused by this. Is this something that might change often?

Reply:
no, but I through habit always give a public member function to change stuff through the constructor. Like you could call with 1 thread to the main constructor, the use case I had in mind was that at the time of loading the DAGMC geometry, one didnt know the number of threads. Some time later, you figure out the number of threads, you can then set the number of threads with that function and then setup accordingly.

Reply:
I see what you mean. There should probably be some way to handle going from N to M threads, deleting or constructing instances as needed, yeah?

The case where M < N is probably not very likely, but tackling the general problem early will mean it's taken care of for good. Could there be a method called update_child_threads which would handle this?

Reply:
TBH I only really imagined from going from 1->M, I hadnt thought about going down, but yes to do it properly one should have an update method.